gpflow/utilities/utilities.py

```diff
@@ -246,22 +247,123 @@ def clear_bijector(bijector, _, state):
     return input_module
 
 
-M = TypeVar('M', bound=tf.Module)
+def _get_by_name_index(parent: object, attr_str: str, index_str: str) -> object:
+    attr = getattr(parent, attr_str)
+    if index_str != "":
+        index = int(index_str)
+        return attr[index]
+    return attr
+
+
+def _set_by_name_index(parent: object, value: Any, attr_str: str, index_str: str):
+    if index_str != "":
+        index = int(index_str)
+        attr = getattr(parent, attr_str)
+        attr[index] = value
+    else:
+        setattr(parent, attr_str, value)
+
+
+def _get_last_attr_spec(parent: object, attr_path: str) -> Tuple[object, str, str]:
+    """Returns second to last attribute, the next attribute name, and
+    an index if there is a list access.
+
+    Example:
+        module = ModuleWithNestedStructure(...)
+        attr_path = "a.b.c[0].d[1000]"
+        value = module.a.b.c[0].d[1000]
+        c0, dname, dindex = _get_last_attr_spec(module, "a.b.c[0].d[1000]")
+        assert c0 is module.a.b.c[0]
+        assert dname == "d"
+        assert dindex == "1000"
+
+    :param parent: A python object with a nested structure.
+    :param attr_path: An attribute path.
 
+    :returns: The value stored in the nested object by the attribute path.
+    """
 
-def deepcopy_components(input_module: M) -> M:
+    # Regexp extracts attribute names and indices if available.
+    # Outer brackets represent repeated pattern inside and contains 3 groups:
+    # - '(\w+)' is a group for attribute name.
+    # - '(\[(-?[\s\d]+)\])' is the index group and has a subgroup '(-?[\s\d]+)', it may or may
+    #   not appear in the pattern
+    # - '(-?[\s\d]+)' is an index without squared brackets
+    # - Last part of the expression reflects that the period delimeter cannot be last in a path.
+    restr = r"((\w+)(\[(-?[\s\d]+)\])?(?<!\.))+"
+    curr_parent = parent
+    attrs = re.findall(restr, attr_path)
+    for _, attr_str, _, index_str in attrs[:-1]:
+        curr_parent = _get_by_name_index(curr_parent, attr_str, index_str)
+    _, attr_str, _, index_str = attrs[-1]
+    return curr_parent, attr_str, index_str
+
+
+def getattr_by_path(target: object, attr_path: str) -> Any:
+    """Get a value of nested attribute by a string path.
+
+    :param target: Root object that contains nested attribute
+    :param attr_path: String type value that represents path to an attribute.
+
+    Example:
+        k = gpflow.kernels.Matern52()
+        m = gpflow.models.GPR(..., kernel=kernel)
+        lengthscale = getattr_by_path(m, "kernel.lengthscale")
+    """
+    try:
+        descendant, attr, index = _get_last_attr_spec(target, attr_path)
+        return _get_by_name_index(descendant, attr, index)
+    except ValueError as error:
+        raise ValueError(f"Cannot get value at path '{attr_path}'\n"
+                         f"Got error: {error}")
+
+
+def setattr_by_path(target: object, attr_path: str, value: Any):
```

Comment:
can be module-private

Reply:
Why private? The utility function for a user.

Reply:
what would be a typical use case?

Reply:
@vdutor mentioned wanting to make use of this for monitoring